packages/hydrogen/src/analytics-manager/AnalyticsProvider.tsx

```diff
@@ -0,0 +1,524 @@
+  callback: (payload: CartViewPayload) => void,
+): void;
+
+function subscribe(
+  event: typeof AnalyticsEvent.SEARCH_VIEWED,
+  callback: (payload: SearchViewPayload) => void,
+): void;
+
+function subscribe(
+  event: typeof AnalyticsEvent.CART_UPDATED,
+  callback: (payload: CartUpdatePayload) => void,
+): void;
+
+function subscribe(
+  event: typeof AnalyticsEvent.PRODUCT_ADD_TO_CART,
+  callback: (payload: CartLineUpdatePayload) => void,
+): void;
+
+function subscribe(
+  event: typeof AnalyticsEvent.PRODUCT_REMOVED_FROM_CART,
+  callback: (payload: CartLineUpdatePayload) => void,
+): void;
+
+function subscribe(
+  event: typeof AnalyticsEvent.CUSTOM_EVENT,
+  callback: (payload: CustomEventPayload) => void,
+): void;
+
+function subscribe(event: any, callback: any) {
+  if (!subscribers.has(event)) {
+    subscribers.set(event, new Map());
+  }
+  subscribers.get(event)?.set(callback.toString(), callback);
+}
+
+const waitForReadyQueue = new Map<any, any>();
+
+function publish(
+  event: typeof AnalyticsEvent.PAGE_VIEWED,
+  payload: PageViewPayload,
+): void;
+function publish(
+  event: typeof AnalyticsEvent.PRODUCT_VIEWED,
+  payload: ProductViewPayload,
+): void;
+function publish(
+  event: typeof AnalyticsEvent.COLLECTION_VIEWED,
+  payload: CollectionViewPayload,
+): void;
+function publish(
+  event: typeof AnalyticsEvent.CART_VIEWED,
+  payload: CartViewPayload,
+): void;
+function publish(
+  event: typeof AnalyticsEvent.CART_UPDATED,
+  payload: CartUpdatePayload,
+): void;
+function publish(
+  event: typeof AnalyticsEvent.PRODUCT_ADD_TO_CART,
+  payload: CartLineUpdatePayload,
+): void;
+function publish(
+  event: typeof AnalyticsEvent.PRODUCT_REMOVED_FROM_CART,
+  payload: CartLineUpdatePayload,
+): void;
+function publish(
+  event: typeof AnalyticsEvent.CUSTOM_EVENT,
+  payload: OtherData,
+): void;
+function publish(event: any, payload: any): void {
+  if (!areRegistersReady()) {
+    waitForReadyQueue.set(event, payload);
+    return;
+  }
+
+  publishEvent(event, payload);
+}
+
+function publishEvent(event: any, payload: any): void {
+  (subscribers.get(event) ?? new Map()).forEach((callback, subscriber) => {
+    try {
+      callback(payload);
+    } catch (error) {
+      if (typeof error === 'object' && error instanceof Error) {
+        console.error(
+          'Analytics publish error',
+          error.message,
+          subscriber,
+          error.stack,
+        );
+      } else {
+        console.error('Analytics publish error', error, subscriber);
+      }
+    }
+  });
+}
+
+function register(key: string) {
+  if (!registers.hasOwnProperty(key)) {
+    registers[key] = false;
+  }
+
+  return {
+    ready: () => {
+      registers[key] = true;
+
+      if (areRegistersReady() && waitForReadyQueue.size > 0) {
+        waitForReadyQueue.forEach((queuePayload, queueEvent) => {
+          publishEvent(queueEvent, queuePayload);
+        });
+        waitForReadyQueue.clear();
+      }
+    },
```

Comment:
Mmm ... I think this would introduce further more complex eventing further down the timeline. We need to discard all events (or stop queueing) after the first soft navigation. If consent is not given, we should not keep any events in queue ... so queueing the events are only for the purpose of from first load until we got the config for consent.